walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,67 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+
+    // Find times that don't work
+    ArrayList<TimeRange> badRanges = new ArrayList<>();
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      for (String attendee : attendees) {
+        if (eventAttendees.contains(attendee)) {
+          badRanges.add(event.getWhen());
+          break;
+        }
+      }
+    }
+    badRanges.sort(TimeRange.ORDER_BY_START);
+
+    // Prune time ranges
+    for (int i = 0; i < badRanges.size() - 1; i++) {
+      TimeRange range1 = badRanges.get(i);
+      TimeRange range2 = badRanges.get(i + 1);
+      // Range2 can contain range1 if they have equal start times, but that case is equivalent to
+      // overlap
+      if (range1.overlaps(range2)) {
+        if (range1.contains(range2)) {
+          // Irrelevant time range
```

Comment:
I don't understand how this code works. If at step 3, you null out range 4, then in the next iteration of the for loop, when i=4, it seems like you are going to null pointer exception on line 43?

Big picture:
  1) I would recommend creating a second arraylist for the nonredundant array lists, rather than modifying the list as you traverse it, since this is quite error prone.

 2) I would restructure your code a bit here. I think you probably want nested loops.

List<TimeRange> filteredRanges;
for(int i =0; i < num_ranges; i++) {
  Range currentUnion = ranges.get(i);
  Range next =ranges.get(i+1);
  while(current_union.contains(next)) {
    currentUnion = currentUnion.combine(next);
     i++;
  }
  filteredRanges.add(currentUnion);
}